File: app/src/main/java/opencrowd/hgc/hgcwallet/ui/onboard/RestoreWalletScreen.java
package opencrowd.hgc.hgcwallet.ui.onboard;

import android.content.Context;
import android.support.annotation.NonNull;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;

import com.wealthfront.magellan.BaseScreenView;
import com.wealthfront.magellan.Screen;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import opencrowd.hgc.hgcwallet.R;
import opencrowd.hgc.hgcwallet.common.Singleton;
import opencrowd.hgc.hgcwallet.common.UserSettings;
import opencrowd.hgc.hgcwallet.crypto.HGCSeed;
import opencrowd.hgc.hgcwallet.modals.HGCKeyType;
import opencrowd.hgc.hgcwallet.ui.customviews.TitleBarWrapper;

public class RestoreWalletScreen extends Screen<RestoreWalletView> {

    @NonNull
    HGCKeyType keyType;

    public RestoreWalletScreen(@NonNull HGCKeyType keyType)
    {
        this.keyType = keyType;
    }

    @NonNull
    @Override
    protected RestoreWalletView createView(@NonNull Context context) {
        return new RestoreWalletView(context,keyType);
    }
}

class RestoreWalletView extends BaseScreenView<RestoreWalletScreen> {

    Button mRestoreButton;
    ImageView mImageView;
    EditText mEdittextRestore;

    public RestoreWalletView(@NonNull Context context, final @NonNull HGCKeyType keyType) {
        super(context);
        inflate(context, R.layout.view_restore_layout, this);
        TitleBarWrapper titleBar = new TitleBarWrapper(findViewById(R.id.titleBar));
        titleBar.setTitle(R.string.restore_wallet_account);

        mImageView = (ImageView)findViewById(R.id.image_close);
        mEdittextRestore = (EditText)findViewById(R.id.edittext_restore);
        mRestoreButton = (Button)findViewById(R.id.btn_restore);
        titleBar.setCloseButtonHidden(false);
        titleBar.setOnCloseButtonClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                getScreen().getNavigator().goBack();
            }
        });

        mRestoreButton.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View view) {
                try {
                    List<String> allWords = Collections
                            .synchronizedList(new ArrayList<String>());
                    Matcher m = Pattern.compile("[a-zA-Z]+")
                            .matcher(mEdittextRestore.getText().toString().toLowerCase());
                    while (m.find()) {
                        allWords.add(m.group());
                    }

                    HGCSeed seed = new HGCSeed(allWords);
                    if (allWords.size() == HGCSeed.bip39WordListSize) {
                        UserSettings.instance.setValue(UserSettings.KEY_HAS_SHOWN_BIP39_MNEMONIC, true);
                    }
                    getScreen().getNavigator().goTo(new RestoreAccountIDScreen(keyType,seed));

                } catch (Exception e) {
                    Singleton.INSTANCE.showToast(getScreen().getActivity(),getScreen().getActivity().getString(R.string.invalid_phrase));
                    e.printStackTrace();
                }
            }
        });

    }
}

File: app/src/main/java/hedera/hgc/hgcwallet/ui/onboard/RestoreWalletScreen.kt
/*
 *
 *  Copyright 2019 Hedera Hashgraph LLC
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */

package hedera.hgc.hgcwallet.ui.onboard

import android.content.Context
import android.view.View
import android.widget.Button
import android.widget.EditText
import com.wealthfront.magellan.BaseScreenView
import com.wealthfront.magellan.Screen
import hedera.hgc.hgcwallet.R
import hedera.hgc.hgcwallet.common.Singleton
import hedera.hgc.hgcwallet.crypto.HGCSeed
import hedera.hgc.hgcwallet.modals.KeyDerivation
import hedera.hgc.hgcwallet.ui.customviews.TitleBarWrapper
import java.util.*
import java.util.regex.Pattern

class RestoreWalletScreen() : Screen<RestoreWalletView>() {

    data class Param(val str: String = "")

    private val param: Param

    init {
        param = Param()
    }

    override fun createView(context: Context): RestoreWalletView {
        return RestoreWalletView(context, param)
    }

    internal fun onCloseButtonClick() {
        navigator?.goBack()
    }

    internal fun onRestoreButtonClick(str: String) {
        try {
            val allWords = Collections
                    .synchronizedList(ArrayList<String>())
            val m = Pattern.compile("[a-zA-Z]+")
                    .matcher(str.toLowerCase())
            while (m.find()) {
                allWords.add(m.group())
            }

            val seed = HGCSeed(allWords)
            var keyDerivation: KeyDerivation? = null
            if (allWords.size != HGCSeed.bip39WordListSize)
                keyDerivation = KeyDerivation.HGC

            navigator?.goTo(RestoreAccountIDScreen(seed, keyDerivation))

        } catch (e: Exception) {
            Singleton.showToast(activity, activity.getString(R.string.invalid_phrase))
            e.printStackTrace()
        }
    }
}

class RestoreWalletView(context: Context, param: RestoreWalletScreen.Param) : BaseScreenView<RestoreWalletScreen>(context) {

    private val edittextRestore: EditText?

    init {
        View.inflate(context, R.layout.view_restore_layout, this)
        TitleBarWrapper(findViewById(R.id.titleBar)).apply {
            setTitle(R.string.restore_wallet_account)
            setCloseButtonHidden(false)
            setOnCloseButtonClickListener {
                screen?.onCloseButtonClick()
            }
        }

        edittextRestore = findViewById<EditText>(R.id.edittext_restore)
        findViewById<Button>(R.id.btn_restore)?.apply {
            setOnClickListener {
                screen?.onRestoreButtonClick(edittextRestore.getText().toString())

            }
        }
    }
}